kratos/includes/registry_item.h

```diff
@@ -0,0 +1,282 @@
+//
+//  Main authors:    Pooyan Dadvand
+//
+
+#if !defined(KRATOS_REGISTRY_ITEM_H_INCLUDED )
+#define  KRATOS_REGISTRY_ITEM_H_INCLUDED
+
+
+// System includes
+#include <string>
+#include <iostream>
+#include <unordered_map>
+
+
+// External includes
+
+
+// Project includes
+#include "includes/define.h"
+
+
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/// The registry item to be stored by Registry class. It is the base class for some more specific ones.
+/** RegistryItem has a tree node structure and stores its name, an optional
+ *  value, and an unorder_set of its sub data. 
+ *  This structure let us to have registry of the elements and then different
+ *  registries for each elements inside it.
+ *  Please note that RegistryItem stores a pointer to the value. 
+ *  To have a copy of the value you may use the derived RegistryValueItem 
+ *  which crates a copy in construction and delete it in its destructor 
+ *  to make the memory management easier.
+*/
+class RegistryItem
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of RegistryItem
+    KRATOS_CLASS_POINTER_DEFINITION(RegistryItem);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor deleted.
+    RegistryItem() = delete;
+
+    /// Constructor with the name
+    RegistryItem(std::string Name) : mName(Name), mpValue(nullptr){}
+
+    /// Constructor with the name and value
+    template<typename TValueType>
+    RegistryItem(std::string Name, TValueType const& Value) : mName(Name), mpValue(&Value){}
+
+    /// Destructor.
+    virtual ~RegistryItem(){
+    }
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Items
+    ///@{
+
+    template< typename TItemType, class... TArgumentsList >
+    RegistryItem& AddItem(std::string const& ItemName, TArgumentsList&&... Arguments){
+        KRATOS_ERROR_IF(this->HasItem(ItemName)) << "The RegistryItem " << this->Name() << " already has an item with name " << ItemName << std::endl;
+        auto insert_result = mSubRegistryItem.emplace(std::make_pair(ItemName, Kratos::make_unique<TItemType>(ItemName, std::forward<TArgumentsList>(Arguments)...)));
+        KRATOS_ERROR_IF_NOT(insert_result.second) << "Error in inserting " << ItemName << " in registry item with name " << this->Name() << std::endl;
+        return *insert_result.first->second;
+    }
+
+
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    const std::string& Name() const{
+        return mName;
+    }
+
+    RegistryItem const& GetItem(std::string const& ItemName) const;
+
+    RegistryItem& GetItem(std::string const& ItemName);
+
+    void RemoveItem(std::string const& ItemName);
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+        bool HasValue() const{
```

Comment:
To many spaces (to avoid many comments I just say here)

Reply:
Done.